### PR TITLE
Maturity level two

### PR DESCRIPTION
**Pull request by pietersap**
Moved to codito car domain. Data model is very similar to original Worldcup sample. One Car has multipleCustomizations, just like one Team has multiple Players.

### CODE REVIEW

File: maturity-level-two/src/Controllers/v1/CustomizationController.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        public async Task<IActionResult> VoteAsBestPlayer(int id)
+        {
+            var customization = await _coditoRepository.GetCustomizationAsync(id);
+            if (customization == null)
+            {
+                return NotFound(new ProblemDetailsError(StatusCodes.Status404NotFound));
+            }
+            if (customization.InventoryLevel <= 0)
+            {
+                return BadRequest();
```
Comment (fgheysels) [inline maturity-level-two/src/Controllers/v1/CustomizationController.cs]: This should not result in a BadRequest response.

A BadRequest indicates that something is wrong at the client-level; something is wrong with the request object.  This is not the case when you have no inventory left.
A NotAllowed statuscode might be more appropriate although I'm not convinced of it either.

File: maturity-level-two/src/Controllers/v1/CustomizationController.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            }
+
+            await _coditoRepository.ApplyCustomizationSaleAsync(customization);
```
Comment (fgheysels) [inline maturity-level-two/src/Controllers/v1/CustomizationController.cs]: Actually, all these operations on the database should be done inside a (exclusive) transaction, since a race-condition is introduced here.

It is possible that multiple sales requests for the same item are coming in at the same time.  User A can retrieve the customization to check the inventory level, and that might be fine but by the time the `ApplyCustomizationSaleAsync` method is executed, the inventory level might already be changed to 0 by other incoming requests.

File: maturity-level-two/src/Controllers/v1/CustomizationController.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+            return Accepted();
+        }
+
+
+
```
Comment (fgheysels) [inline maturity-level-two/src/Controllers/v1/CustomizationController.cs]: Unnecessary whitespace :)

File: maturity-level-two/src/Entities/Car.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+﻿using System.ComponentModel.DataAnnotations;
+using System.Collections.Generic;
+
+
+namespace Codit.LevelTwo.Entities
+{
+    public class Car
+    {
+        [Key]
+        public int Id { get; set; }
```
Comment (fgheysels) [inline maturity-level-two/src/Entities/Car.cs]: IMHO, an Id should not be settable, so the setter should be private.

File: maturity-level-two/src/Entities/Car.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+        [Required]
+        public string Model { get; set; }
+
+        //public string BodyType { get; set; }
+        public CarBodyType BodyType { get; set; }
+
+        public string Description { get; set; }
+
+        public ICollection<Customization> Customizations { get; set; } = new List<Customization>();
```
Comment (fgheysels) [inline maturity-level-two/src/Entities/Car.cs]: This property should be readonly

```csharp
public ICollection<Customization> Customizations { get;  } = new List<Customization>();
```

File: maturity-level-two/src/Entities/Customization.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
+﻿using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+
+
+
```
Comment (fgheysels) [inline maturity-level-two/src/Entities/Customization.cs]: Unnecessary whitespace.

File: maturity-level-two/src/Entities/Customization.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+namespace Codit.LevelTwo.Entities
+{
+    public class Customization
+    {
+        [Key]
+        public int Id { get; set; }
```
Comment (fgheysels) [inline maturity-level-two/src/Entities/Customization.cs]: Same remark about settable id's.

File: maturity-level-two/src/Extensions/ApplicationBuilderExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                    var problemDetails = new ProblemDetailsError
+                    {
+                        Title = "An unexpected error occurred!",
+                        Status = StatusCodes.Status500InternalServerError,
+                        Detail = errorDetail,
+                        Instance = $"urn:codit.eu:server-error:{Guid.NewGuid()}"
+                    };
+
+                    // TODO: Plug in telemetry
+                    context.Response.WriteJson(problemDetails, contentType: ContentTypeNames.Application.JsonProblem);
```
Comment (fgheysels) [inline maturity-level-two/src/Extensions/ApplicationBuilderExtensions.cs]: Shouldn't we let ASP.NET handle the serialization of the error in the format that is specified by the accept header of the request instead of deciding for ourselves that it should always be json ?

File: maturity-level-two/src/Services/CoditoRepository.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+using Microsoft.EntityFrameworkCore;
+
+
+
+namespace Codit.LevelTwo.Services
+{
+    public class CoditoRepository : ICoditoRepository
+    {
+
+        public CoditoContext _coditoContext;
```
Comment (fgheysels) [inline maturity-level-two/src/Services/CoditoRepository.cs]: `private readonly` please.

File: maturity-level-two/src/Services/CoditoRepository.cs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            }
+
+            return await _coditoContext.Cars.Where(car => car.Id == id).FirstOrDefaultAsync();
+        }
+
+        public async Task<IEnumerable<Customization>> GetAllCustomizationsAsync()
+        {
+
+           return await _coditoContext.Customizations.OrderByDescending(cust => cust.NumberSold).ToListAsync();
+
```
Comment (fgheysels) [inline maturity-level-two/src/Services/CoditoRepository.cs]: Unnecessary whitespace above and below the single line of code.

File: maturity-level-two/src/Services/CoditoRepository.cs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        public async Task CreateCustomizationAsync(Customization customization)
+        {
+
+             _coditoContext.Customizations. AddAsync(customization);
+            await _coditoContext.SaveChangesAsync();
+            
+        }
+
+        public async Task ApplyPatchAsync<TEntity, TDto>(TEntity entityToUpdate, TDto dto) where TEntity : class
```
Comment (fgheysels) [inline maturity-level-two/src/Services/CoditoRepository.cs]: The `TDto` represents a model that is exposed by the API.  I wouldn't use that as a parameter in the repository.
What I would do, is pass in the `TEntity` in where the properties are already modified.

File: maturity-level-two/src/Services/CoditoRepository.cs [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+            }
+
+            await _coditoContext.SaveChangesAsync();
+        }
+
+        public async Task DeleteCustomizationAsync(int id)
+        {
+            var entityToDelete = _coditoContext.Customizations.Find(id);
+            _coditoContext.Customizations.Remove(entityToDelete);
+            await _coditoContext.SaveChangesAsync();
```
Comment (fgheysels) [inline maturity-level-two/src/Services/CoditoRepository.cs]: We shouldn't delete items in this way.

The code above will generate a SQL query that will first retrieve the entity, and then you'll use that entity to delete it.  This is not a good solution performance-wise as you'll end up with an unnecessary roundtrip to the DB.

A better solution would be something like this:

```csharp
var entityToDelete = new Customization(id);
_coditoContext.Customizations.Entry[entityToDelete].State = EntityState.Deleted;
await _coditoContext.SaveChangesAsync();
```
or
```csharp
var entityToDelete = new Customization(id);
_coditoContext.Customizations.Attach(entityToDelete);
_coditoContext.Customizations.Delete(entityToDelete);
await _coditoContext.SaveChangesAsync();
```

File: maturity-level-two/src/Services/CoditoRepository.cs [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        public async Task ApplyCustomizationSaleAsync(Customization customization)
+        {
+            //var entity = await _coditoContext.Customizations.Where(cust => cust.Id == id).FirstOrDefaultAsync();
+            customization.InventoryLevel --;
+            customization.NumberSold++;
```
Comment (fgheysels) [inline maturity-level-two/src/Services/CoditoRepository.cs]: IMO, there should be a method call `Sell` on the `Customization` type which decreases InventoryLevel and increases NumberSold atomically.  This would be a better way to encapsulate the logic that you want to expose with your model and prevents that someone increases NumberSold without decreasing the inventorylevel

File: maturity-level-two/src/Services/CoditoRepository.cs [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        public async Task DeleteCustomizationAsync(int id)
+        {
+            var entityToDelete = _coditoContext.Customizations.Find(id);
+            _coditoContext.Customizations.Remove(entityToDelete);
+            await _coditoContext.SaveChangesAsync();
+        }
+
+        public async Task ApplyCustomizationSaleAsync(Customization customization)
+        {
+            //var entity = await _coditoContext.Customizations.Where(cust => cust.Id == id).FirstOrDefaultAsync();
```
Comment (fgheysels) [inline maturity-level-two/src/Services/CoditoRepository.cs]: Commented code should be removed

File: maturity-level-two/tests/Codit.IntegrationTest/CarTest.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        private readonly HttpClient _httpClient;
+        public CarTest()
+        {
+            if (_httpClient != null) return;
+            var srv = new TestServer(new WebHostBuilder()
+                .UseEnvironment("Development")
+                .UseStartup<Startup>());
+
+            _httpClient = srv.CreateClient();
```
Comment (fgheysels) [inline maturity-level-two/tests/Codit.IntegrationTest/CarTest.cs]: For each testcase that is executed, you're creating a new HttpClient instance.

It would be better to have it like this:
```csharp
private static readonly HttpClient Http;

static CarTest()
{
   var srv = new TestServer(.....);

   Http = srv.CreateClient();
}
```

File: maturity-level-two/tests/Codit.IntegrationTest/CustomizationTest.cs [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+    {
+        private readonly HttpClient _httpClient;
+        public CustomizationTest()
+        {
+            if (_httpClient != null) return;
+            var srv = new TestServer(new WebHostBuilder()
+                .UseEnvironment("Development")
+                .UseStartup<Startup>());
+
+            _httpClient = srv.CreateClient();
```
Comment (fgheysels) [inline maturity-level-two/tests/Codit.IntegrationTest/CustomizationTest.cs]: Same remark as in the CarTest file.

File: maturity-level-two/tests/Codit.IntegrationTest/CustomizationTest.cs [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+            response.StatusCode.Should().Be(HttpStatusCode.NotFound);
+
+        }
+
+        [Fact]
+        public async Task DeleteCustomization_NotFound_TestAsync()
+        {
+            //Arrange
+            int id = 1;
+            var request = new HttpRequestMessage(new HttpMethod("DELETE"), $"/codito/v1/customization/{id}");
```
Comment (fgheysels) [inline maturity-level-two/tests/Codit.IntegrationTest/CustomizationTest.cs]: Why not

```csharp
new HttpRequestMessage(HttpMethod.Delete, ...);
```

File: maturity-level-two/tests/Codit.UnitTest/Controllers/carControllerTest.cs [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+namespace Codit.UnitTest.Controllers
+{
+    public class CarControllerTest
+    {
+        private readonly CarController _controller;
+
+        public CarControllerTest()
+        {
+            _controller = new CarController(new CoditoRepositoryFake());
+            AutoMapperConfig.Initialize();
```
Comment (fgheysels) [inline maturity-level-two/tests/Codit.UnitTest/Controllers/carControllerTest.cs]: Automapper should be initialized only once during the lifetime of the application, so it should not be called in the ctor of a test.  
You could create an xunit fixture that does this.

File: maturity-level-two/src/Services/ICoditoRepository.cs [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+        Task<Customization> GetCustomizationAsync(int id);
+
+        Task CreateCustomizationAsync(Customization customization);
+        Task ApplyPatchAsync<TEntity, TDto>(TEntity entityToUpdate, TDto dto) where TEntity : class;
+
+        Task ApplyCustomizationSaleAsync(Customization customization);
+
+        Task DeleteCustomizationAsync(int id);
+
+
```
Comment (fgheysels) [inline maturity-level-two/src/Services/ICoditoRepository.cs]: Unnecessary whitespace

File: maturity-level-two/tests/Codit.IntegrationTest/CarTest.cs [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+            //Arrange
+            var request = new HttpRequestMessage(new HttpMethod("GET"), "/codito/v1/car");
+            //Act
+            var response = await _httpClient.SendAsync(request);
+            //Assert
+            response.StatusCode.Should().Be(HttpStatusCode.OK);
+        }
+
+        [Fact]
+        public async Task GetSingleTeam_Ok_TestAsync()
```
Comment (fgheysels) [inline maturity-level-two/tests/Codit.IntegrationTest/CarTest.cs]: Still refering to the old model

File: maturity-level-two/tests/Codit.IntegrationTest/CarTest.cs [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+            int id = 1;
+            var request = new HttpRequestMessage(new HttpMethod("GET"), $"/codito/v1/car/{id}");
+            //Act
+            var response = await _httpClient.SendAsync(request);
+            //Arrange
+            response.StatusCode.Should().Be(HttpStatusCode.OK);
+        }
+
+        [Fact]
+        public async Task GetSingleTeam_NotFound_TestAsync()
```
Comment (fgheysels) [inline maturity-level-two/tests/Codit.IntegrationTest/CarTest.cs]: Still referring to the old model.

File: maturity-level-two/src/Services/ICoditoRepository.cs [hunk 22]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+using Codit.LevelTwo.Entities;
+
+namespace Codit.LevelTwo.Services
+{
+    public interface ICoditoRepository
```
Comment (fgheysels) [inline maturity-level-two/src/Services/ICoditoRepository.cs]: You should have a repository per aggregate root, and not one global repository.
I think in this domain model, we currently have only one aggregate root, which would be `Car`

IMHO it makes no sense to have a method `GetAllCustomizations`; you'll rather retrieve the Customizations that are available per `CarBodyType` instead.

Actually, I'd create this model in a whole other fashion, but I do not know how 'free' you are to do so.  I'd suggest to have an entity `Car` which has a property `Brand` and a property that reflects its type.   That property should not be an entity itself as it is a value-type.  
But that might be outside the scope of this project :)

File: maturity-level-two/README.md [hunk 23]
```diff
@@ -0,0 +1,201 @@
+# Practical API Guidelines - Must have
```
[the remark this answers is not in the document]
Comment (pietersap) [reply maturity-level-two/README.md]: We could remove them from both subfolders and just add a "GUIDELINES.md" to the root folder

File: maturity-level-two/src/Constants.cs [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            public const string Title = "Codito Car API";
+            public const string Description = "Codito Car models and customizations";
+            public const string TermsOfService = "N/A";
+            public const string ContactName = "API at Codit";
+            public const string ContactEmail = "support@codit.eu";
+            public const string ContactUrl = "https://www.codit.eu";
+        }
+
+        public static class RouteNames
```
[the remark this answers is not in the document]
Comment (pietersap) [reply maturity-level-two/src/Constants.cs]: They are also used in the integration test (OpenApiTest.cs)

File: maturity-level-two/src/Controllers/v1/CustomizationController.cs [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+        /// Sell a car customization
+        /// </summary>
+        /// <param name="id">Customization identifier</param>
+        /// <returns>Acknowledge that the c</returns>
+        [HttpPost("{id}/sale", Name = Constants.RouteNames.v1.SellCustomization)]
+        [SwaggerResponse((int)HttpStatusCode.Accepted, "Sale accepted. No response body")]
+        [SwaggerResponse((int)HttpStatusCode.BadRequest, "Out of stock")]
+        [SwaggerResponse((int)HttpStatusCode.NotFound, "Car customization not found")]
+        [SwaggerResponse((int)HttpStatusCode.InternalServerError, "API is not available")]
+        public async Task<IActionResult> VoteAsBestPlayer(int id)
```
[the remark this answers is not in the document]
Comment (pietersap) [reply maturity-level-two/src/Controllers/v1/CustomizationController.cs]: Copy-paste adventures...

File: maturity-level-two/src/Controllers/v1/CustomizationController.cs (hunk 1 above)
Comment (pietersap) [reply maturity-level-two/src/Controllers/v1/CustomizationController.cs]: @MassimoC We discussed this... NotAllowed is okay for me

File: maturity-level-two/README.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (tomkerkhove) [reply maturity-level-two/README.md]: As a consumer I would prefer to have seperate guidelines, but only list the ones that are applicable to level 2 here.

[the remark this answers is not in the document]
Comment (MassimoC) [reply maturity-level-two/README.md]: The readme.md file in the second maturity level should only contains the guidelines for the second level. The practical implementation will be incremental (level 1+ level 2)

File: maturity-level-two/src/Controllers/v1/CarController.cs [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+    [ApiController]
+    [ValidateModel]
+    public class CarController : ControllerBase
+    {
+        private readonly ICoditoRepository _coditoRepository;
+
+        public CarController(ICoditoRepository coditoRepository)
+        {
+            _coditoRepository = coditoRepository;
+            AutoMapperConfig.Initialize();
```
[the remark this answers is not in the document]
Comment (MassimoC) [reply maturity-level-two/src/Controllers/v1/CarController.cs]: @fgheysels you are right. I will create a separate issue for this one.

File: maturity-level-two/src/Constants.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (MassimoC) [reply maturity-level-two/src/Constants.cs]: I will create a separate issue for this one.

File: maturity-level-two/src/Controllers/v1/CarController.cs (hunk 26 above)
[the remark this answers is not in the document]
Comment (MassimoC) [reply maturity-level-two/src/Controllers/v1/CarController.cs]: Marked as resolved - we will pick this up later as separate issue

File: maturity-level-two/src/Controllers/v1/CarController.cs [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        /// <summary>
+        /// Get all cars
+        /// </summary>
+        /// /// <param name="bodyType">Filter a specific body Type (optional)</param>
+        /// <remarks>Get all cars</remarks>
+        /// <returns>List of cars</returns>
+        [HttpGet(Name = Constants.RouteNames.v1.GetCars)]
+        [SwaggerResponse((int)HttpStatusCode.OK, "List of Cars")]
```
[the remark this answers is not in the document]
Comment (MassimoC) [reply maturity-level-two/src/Controllers/v1/CarController.cs]: @fgheysels  this will be part of the content negotiation guidelines. For now i mark as resolved, we will pick this up with the classic procedure :  Guideline > review > implementation

File: maturity-level-two/src/Extensions/ApplicationBuilderExtensions.cs (hunk 8 above)
Comment (MassimoC) [reply maturity-level-two/src/Extensions/ApplicationBuilderExtensions.cs]: @fgheysels  adding this comment to the Content-Negotiation issue

File: maturity-level-two/src/Services/CoditoRepository.cs (hunk 11 above)
Comment (MassimoC) [reply maturity-level-two/src/Services/CoditoRepository.cs]: issue created

File: maturity-level-two/src/Controllers/v1/CustomizationController.cs [hunk 28]
```diff
@@ -0,0 +1,186 @@
...
+        [SwaggerResponse((int)HttpStatusCode.NotFound, "Customization not found")]
+        [SwaggerResponse((int)HttpStatusCode.InternalServerError, "API is not available")]
+        public async Task<IActionResult> DeleteCustomization(int id)
+        {
+            var CustomizationObj = await _coditoRepository.GetCustomizationAsync(id);
+            if (CustomizationObj == null)
+            {
+                return NotFound(new ProblemDetailsError(StatusCodes.Status404NotFound));
+            }
+            await _coditoRepository.DeleteCustomizationAsync(id);
```
[the remark this answers is not in the document]
Comment (MassimoC) [reply maturity-level-two/src/Controllers/v1/CustomizationController.cs]: issue created

File: maturity-level-two/src/Controllers/v1/CustomizationController.cs [hunk 29]
```diff
@@ -0,0 +1,186 @@
...
+        [SwaggerResponse((int)HttpStatusCode.NoContent, "The data has been updated")]
+        [SwaggerResponse((int)HttpStatusCode.NotFound, "Customization not found")]
+        [SwaggerResponse((int)HttpStatusCode.InternalServerError, "API is not available")]
+        public async Task<IActionResult> UpdateIncremental(int id, [FromBody] CustomizationDto customizationDto)
+        {
+            var CustomizationObj = await _coditoRepository.GetCustomizationAsync(id);
+            if (CustomizationObj == null)
+            {
+                return NotFound(new ProblemDetailsError(StatusCodes.Status404NotFound));
+            }
```
[the remark this answers is not in the document]
Comment (MassimoC) [reply maturity-level-two/src/Controllers/v1/CustomizationController.cs]: issue  created

File: maturity-level-two/src/Controllers/v1/CustomizationController.cs (hunk 1 above)
Comment (MassimoC) [reply maturity-level-two/src/Controllers/v1/CustomizationController.cs]: I created an issue for this. Better to discuss this and make it clear as guideline

File: maturity-level-two/src/Controllers/v1/CustomizationController.cs (hunk 2 above)
Comment (MassimoC) [reply maturity-level-two/src/Controllers/v1/CustomizationController.cs]: issue created

File: maturity-level-two/src/Services/CoditoRepository.cs (hunk 9 above)
Comment (MassimoC) [reply maturity-level-two/src/Services/CoditoRepository.cs]: fixed

**Comment by pietersap** [comment]
I'm working on the enhancements and will commit later
